src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Semantic/DefaultRazorSemanticTokenInfoServiceTest.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Completion;
+using Microsoft.AspNetCore.Razor.LanguageServer.Semantic;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Test.Semantic
+{
+    public class DefaultRazorSemanticTokenInfoServiceTest : DefaultTagHelperServiceTestBase
+    {
+        public static IEnumerable<object[]> TestCases =>
+            new List<object[]>
+            {
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1></test1>",
+                new List<uint> {
+                    1, 1, 5, 0, 0, //line, character pos, length, tokenType, modifier
+                    0, 8, 5, 0, 0
+                }},
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1 bool-val='true'></test1>" ,
+                new List<uint> {
+                    1, 1, 5, 0, 0, //line, character pos, length, tokenType, modifier
+                    0, 6, 8, 1, 0,
+                    0, 18, 5, 0, 0
+                }},
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1 bool-val></test1>",
+                new List<uint> {
+                    1, 1, 5, 0, 0, //line, character pos, length, tokenType, modifier
+                    0, 6, 8, 1, 0,
+                    0, 11, 5, 0, 0
+                }},
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1 bool-val='true' class='display:none'></test1>",
+                new List<uint> {
+                    1, 1, 5, 0, 0, //line, character pos, length, tokenType, modifier
+                    0, 6, 8, 1, 0,
+                    0, 39, 5, 0, 0
+                }},
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<p bool-val='true'></p>",
+                new List<uint> {}}
+            };
+
+        public static IEnumerable<object[]> DirectiveCases =>
+            new List<object[]>
+            {
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1 @onclick='Function'></test1>",
+                new List<uint> {
+                    1, 1, 5, 1, 0, //line, character pos, length, tokenType, modifier
+                    0, 8, 5, 2, 0
+                }},
+                new object[] { $"@addTagHelper *, TestAssembly{Environment.NewLine}<test1 @onclick:preventDefault='Function'></test1>",
+                new List<uint> {
+                    1, 1, 5, 1, 0, //line, character pos, length, tokenType, modifier
+                    0, 8, 5, 2, 0
+                }},
+            };
+
+        [Theory]
+        [MemberData(nameof(TestCases))]
```

Comment:
I'd avoid using `MemberData` for these. Very rarely does `MemberData` make sense. Reasons are that when a test fails it's nearly impossible to re-run that individual test (you have to run all test cases) and detecting which test-case actually fails is hard to do as well because you need to reverse-engineer the test to understand which case failed and then why that case failed.

Versus, if you copy and paste the setup or just make the common setup pieces abstract into methods or properties you can end up with more debuggable/readable tests such as:
```
GetSemanticTokens_TagHelperElement
GetSemanticTokens_TagHelperAttribute
GetSemanticTokens_TagHelperElement_KnownHtmlElement
```

Same thing with the directive cases.